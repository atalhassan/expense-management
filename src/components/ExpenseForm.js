import React from 'react'
import moment from 'moment'
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';


class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount /100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }
  changeByName = (e) => {
    const change = {}
    const name = e.target.name
    change[name] = e.target.value
    if (name === 'amount') {
      const amount = e.target.value
      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
        change[name] = amount
      } else {
        change[name] = this.state.amount
      }
      this.setState(() => change)
    } else {
      this.setState(() => change)
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}))
    }
  }
  onFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}))
  }
  onSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'please fill the description and amount'}))
    } else {
      this.setState(() => ({error: ''}))
      this.props.onSubmit({description: this.state.description,
        amount:parseFloat(this.state.amount, 10) * 100,
        createdAt:this.state.createdAt.valueOf(),
        note: this.state.note})
      }
  }

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <input
            type='text'
            name='description'
            placeholder='Description'
            value={this.state.description}
            onChange={this.changeByName}
            autoFocus
          />
          <input
            type='text'
            name='amount'
            placeholder='amount'
            value={this.state.amount}
            onChange={this.changeByName}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder='Add a note for your expense (optional)'
            name='note'
            value={this.state.note}
            onChange={this.changeByName}
          >
          </textarea>
          {this.state.error && <p>{this.state.error}</p>}
          <button>Add Expense</button>
        </form>
      </div>
    )
  }

}


export default ExpenseForm
