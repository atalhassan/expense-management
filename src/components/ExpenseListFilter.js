import React from 'react'
import {connect} from 'react-redux'
import {setTextFilter, SetEndDate, SetStartDate, SortByAmount, SortByDate} from '../actions/filters'

// import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilter extends React.Component {
  state = {
    calendarFocused: null
  }
  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({calendarFocused}))
  }
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.SortByDate()
    } else if (e.target.value === 'amount') {
      this.props.SortByAmount()
    }
  }
  render() {
    return (
      <div className='content-container'>
        <div className='input_group'>
          <div className='input_group__item'>
            <input
              type='text'
              className='text-input'
              value={this.props.filters.text}
              placeholder='Search expenses'
              onChange={this.onTextChange}
            />

          </div>
          <div className='input_group__item'>
            <select
              className='select'
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
              >
                <option value='date'>Date</option>
                <option value='amount'>Amount</option>
              </select>
            </div>
            <div className='input_group__item'>
              <DateRangePicker
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
              />
            </div>
          </div>
        </div>
      )
    }
  }


  const mapStateToProps = (state) => {
    return {
      filters: state.filters
    }
  }

  const mapDispatchToProps = (dispatch) => {
    return {
      setTextFilter : (text) => dispatch(setTextFilter(text)),
      SortByDate : () => dispatch(SortByDate()),
      SortByAmount: () => dispatch(SortByAmount()),
      setStartDate: (startDate) => dispatch(SetStartDate(startDate)),
      setEndDate: (endDate) => dispatch(SetEndDate(endDate))
    }
  }

  export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
