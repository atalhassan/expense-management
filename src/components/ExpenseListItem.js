import React from 'react'
import {Link} from 'react-router-dom'
import moment from 'moment'
import {connect} from 'react-redux'

export const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>${parseFloat(amount/100, 10)} - {moment(createdAt).format('MMM Do, YYYY')}</p>

  </div>
);



export default connect()(ExpenseListItem)
