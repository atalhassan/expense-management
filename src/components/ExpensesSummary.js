import React from 'react'
import numeral from 'numeral'
import {connect} from 'react-redux'
import getVisibleExpenses from '../selectors/expenses'
import getVisibleExpensesTotal from '../selectors/expense-total'

export const ExpensesSummary = ({expenseCount, expenseTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
  const formattedExpenseTotal = numeral(expenseTotal/100).format('$0,0.00')
  return (
    <div>
      <h1>Viewing {expenseCount} {expenseWord} totalling {formattedExpenseTotal}</h1>
    </div>
  )
}

const mapStateToProps = (state) => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)
  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: getVisibleExpensesTotal(visibleExpenses)
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
