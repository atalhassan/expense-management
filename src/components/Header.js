import React from 'react';
import {NavLink} from 'react-router-dom'

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to='/' activeClassName='is-action' exact>Dashboard </NavLink>
    <NavLink to='/create' activeClassName='is-action' >Create </NavLink>

  </header>
)

export default Header;
