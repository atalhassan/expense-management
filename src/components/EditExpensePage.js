import React from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {removeExpenses, editExpenses} from '../actions/expenses'

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpenses(this.props.expense.id,{...expense})
    this.props.history.push('/')
  }
  onClick = () => {
    this.props.removeExpenses({id: this.props.expense.id})
    this.props.history.push('/')
  }

  render()  {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onClick}
          >Remove</button>
      </div>
    )
  }
}

const mapStateToProp = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpenses: (expense) => dispatch(editExpenses(props.expense.id,{...expense})),
    removeExpenses: () => dispatch(removeExpenses({id: props.expense.id}))
  }
}

export default connect(mapStateToProp, mapDispatchToProps)(EditExpensePage);
