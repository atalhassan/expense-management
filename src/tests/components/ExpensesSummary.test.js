import React from 'react'
import {shallow} from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'

test('should render ExpensesSummary correctly for 1 expense',() => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={2345}/>)
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary correctly for multiple expenses',() => {
  const wrapper = shallow(<ExpensesSummary expenseCount={23} expenseTotal={223456345}/>)
  expect(wrapper).toMatchSnapshot()
})
