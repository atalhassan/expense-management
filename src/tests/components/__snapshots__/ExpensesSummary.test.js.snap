// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly for 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $23.45
  </h1>
</div>
`;

exports[`should render ExpensesSummary correctly for multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $2,234,563.45
  </h1>
</div>
`;
